docs/architecture.md

```diff
@@ -0,0 +1,144 @@
+  on containerd using the
+  [Container Runtime Interface (CRI)](https://github.com/kubernetes/cri-api).
+  The method includes the field `SeccompProfilePath` with the value
+  `/var/lib/kubelet/seccomp/foo.json`.
+- Containerd reads the file `/var/lib/kubelet/seccomp/foo.json`:
+  ```
+  {
+    "listenerPath": "/run/seccomp-agent.socket",
+    "defaultAction": "SCMP_ACT_ALLOW",
+    "syscalls": [
+      {
+        "names": ["mount"],
+        "action": "SCMP_ACT_NOTIFY"
+      }
+  }
+  ```
+- Containerd prepares config.json that includes the file foo.json from above:
+  ```
+  {
+    "linux": {
+      "seccomp": {
+        "listenerPath": "/run/seccomp-agent.socket",
+        "defaultAction": "SCMP_ACT_ALLOW",
+        "syscalls": [
+          {
+            "names": ["mount"],
+            "action": "SCMP_ACT_NOTIFY"
+          }
+  ```
+- Containerd starts runc with `config.json`.
+- Runc gets the seccomp file descriptor by calling the seccomp() system call
+  via libseccomp-golang.
+- Runc sends the seccomp file descriptor to the seccomp agent, along with
+  following
+  [Container Process State](https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#containerprocessstate):
+  ```
+  {
+    "ociVersion": "0.2.0",
+    "fds": ["seccompFd"],
+    "pid": 4422,
+    "metadata": "",
+    "state": {
+      "ociVersion": "0.2.0",
+      "id": "8eea9f6d...",
+      "status": "creating",
+      "pid": 4422,
+      "bundle": "rootfs",
+      "annotations": {}
+    }
+  }
+  ```
+- At this point, the Seccomp Agent has the seccomp file descriptor and the
+  Container Process State contains the following
+  [annotations](https://github.com/containerd/containerd/blob/v1.5.2/pkg/cri/annotations/annotations.go)
+  to identify the pod. Additional information such as the ServiceAccount could
+  be queried to the Kubernetes API.
+  - Pod namespace: `io.kubernetes.cri.sandbox-namespace`
+  - Pod name: `io.kubernetes.cri.sandbox-name`
+  - Container name: `io.kubernetes.cri.container-name`
+
+## Example of system call performed on behalf of the process
+
+![Mount syscall with Seccomp Notify](example-syscall-mount.png "Mount syscall with Seccomp Notify")
+
+Steps:
+- A process executes the mount system call:
+  ```
+  mount("proc", "/proc", "proc", 0, NULL)
+  ```
+- The seccomp subssytem in the kernel executes the BPF program for the
+  seccomp filter.
+  The BPF program receives a `struct seccomp_data` as argument, containing:
+  ```
+  struct seccomp_data {
+    int nr;   // syscall mount = 165
+    u32 arch; // x86_64
+    u64 instruction_pointer;
+    u64 args[6]; // syscall args
+  }
+  ```
+- The BPF program returns `SCMP_ACT_NOTIFY`, meaning the decision is deferred
+  to a seccomp agent.
+- The seccomp agent receives a request from seccomp by running the
+  `SECCOMP_IOCTL_NOTIF_RECV` ioctl on the seccomp file descriptor. The request
```

Comment:
The text is clear. But I find it confusing that in the picture the arrow goes from the kernel to the agent. Like, the kernel never calls the agent, but the syscall from the agent just resturns something.

Maybe have the arrow go from the agent to the kernel with the ioctl details, and then an arrow back with the `struct seccomp_notif` can be clearer?

Reply:
I changed it a bit